indra/newview/llappviewer.cpp

```diff
@@ -1375,6 +1378,51 @@ bool LLAppViewer::frame()
 	return ret;
 }
 
+
+// static
+bool packGameControlInput(LLMessageSystem* msg)
+{
+    if (!LLGameControl::hasInput())
+    {
+        return false;
+    }
+    if (!gSavedSettings.getBOOL("EnableGameControlInput"))
+    {
+        LLGameControl::clearInput();
+        return false;
+    }
+
+    msg->newMessageFast(_PREHASH_GameControlInput);
+    msg->nextBlock("AgentData");
+    msg->addUUID("AgentID", gAgentID);
+    msg->addUUID("SessionID", gAgentSessionID);
+
+    const LLGameControl::State& state = LLGameControl::getState();
+
+    size_t num_indices = state.mAxes.size();
+    for (U8 i = 0; i < num_indices; ++i)
+    {
+        if (state.mAxes[i] != state.mPrevAxes[i])
+        {
+            // only pack an axis if it differes from previously packed value
+            msg->nextBlockFast(_PREHASH_AxisData);
+            msg->addU8Fast(_PREHASH_Index, i);
+            msg->addS16Fast(_PREHASH_Value, state.mAxes[i]);
+        }
+    }
+
+    const std::vector<U8>& buttons = state.mPressedButtons;
+    if (!buttons.empty())
+    {
+        msg->nextBlockFast(_PREHASH_ButtonData);
+        msg->addBinaryDataFast(_PREHASH_Data, (void*)(buttons.data()), (S32)(buttons.size()));
+    }
+
+    LLGameControl::clearInput();
```

Comment:
so if I understand it corretly it woll be something like leyboard | joystick -> LLGameControl -> server?

What if user wants some local function on the joystick, like toggling on voice?

Reply:
That is a good question.  I will think about it more when we get closer to something that is shippable, however I don't think mapping game controller input to viewer-only events should be included in a "minimal viable feature" requirement.  If most game controllers are detected, can be remapped on a per-device basis, and the data can reach the server... then we'll probably ship it in that working state and then add bells and whistles in separate PRs.

Reply:
>If most game controllers are detected, can be remapped on a per-device basis, and the data can reach the server...

That feel like doing things in wrong order. Viewer already has basic controller/joystick support, why not hook that into server messages and separately hook controllers into existing UI? Ability to hook axises and more butons to scripts will be very welcome by users even without having better joystick support. And having two separate joystick implementations is going to cause confusion.